nixos/lib/test-driver/test-driver.py

```diff
@@ -1133,12 +1142,14 @@ def cmd(scripts: List[str]) -> Iterator[NixStartScript]:
             for s in scripts:
                 yield NixStartScript(s)
 
+        self.failure_modes = []
         self.machines = [
             Machine(
                 start_command=cmd,
                 keep_vm_state=keep_vm_state,
                 name=cmd.machine_name,
                 tmp_dir=tmp_dir,
+                driver=self,
```

Comment:
hm... i am a bit unhappy with the machine knowing about the driver, and driver-knowledge becoming part of the machine interface.

can we maybe discuss if throwing exceptions from the driver at the machines in case something goes bad (like vscodium quitting)?

Reply:
I agree... Maybe split out the new logic to a class that both `Machine` and `Driver` hold a reference to?

Reply:
not sure how much the machine actually *shall* know about this logic.

in the case of your vscodium example, we could have a thread running that monitors if vscodium is running at all. and as soon as it is not, it could throw an exception at the main thread and we would break out of any check we are in in that moment.

```python
    if fail_early():
        raise Exception("action interrupted early")
```

... as you are already throwing an exception in this case, it could simply be an asynchronous one that is thrown from outside. the advantages in my eyes are:

- `machine` does not have to know or bother about early fails, but fails early when receiving the exception: in *any* code path, not only the ones where you added fail_early support.
- our early fail could even happen during blocking calls (in contrast, in the current implementation we have to wait for a return to then re-check the early-fail predicate)
- you can decide how often/how/when your check condition is called and how complicated it is

Reply:
This was the first approach I tried to take. However, it's more complicated to throw exceptions between threads than I thought... I couldn't get it to work, not sure if it's actually possible to do this in a nice way.

Reply:
(Or maybe we could just pass the bound method `fail_early` of `Driver` instead of the entire `Driver` class?)